tests/backend_test.py

```diff
@@ -1430,3 +1432,69 @@ def test_ibmq_local_emulator(
     counts = r.get_counts()
     # Most results should be (0,0) or (1,1):
     assert sum(c0 != c1 for c0, c1 in counts) < 25
+
+
+# https://github.com/CQCL/pytket-qiskit/issues/231
+def test_noiseless_density_matrix_simulation() -> None:
+    density_matrix_backend = AerDensityMatrixBackend()
+    assert density_matrix_backend.supports_density_matrix is True
+
+    assert isinstance(density_matrix_backend.backend_info.architecture, FullyConnected)
+
+    circ1 = Circuit(3).X(0).X(1).CCX(0, 1, 2)
+
+    output_state = np.array([0] * 7 + [1])
+
+    result1 = density_matrix_backend.run_circuit(circ1)
+    noiseless_dm1 = result1.get_density_matrix()
+
+    assert noiseless_dm1.shape == (8, 8)
+    assert np.allclose(noiseless_dm1, np.outer(output_state, output_state.conj()))
+    # Check purity to verify that we have a pure state
+    assert np.trace(noiseless_dm1**2).real == 1
+
+    # Example with resets and conditional gates
+    # Prepares a state deterministically if input is a computational basis state
+    circ2 = Circuit(3, 1)
+    circ2.CCX(*range(3))
+    circ2.U1(1 / 4, 2)
+    circ2.H(2)
+    circ2.Measure(2, 0)
+    circ2.CZ(0, 1, condition_bits=[0], condition_value=1)
+    circ2.Reset(2)
+
+    result2 = density_matrix_backend.run_circuit(circ2)
+    assert result1.get_density_matrix().shape == (8, 8)
+    state_backend = AerStateBackend()
+    statevector = state_backend.run_circuit(circ2).get_state()
+    noiseless_dm2 = result2.get_density_matrix()
+    assert np.allclose(noiseless_dm2, np.outer(statevector, statevector.conj()))
+    # Check purity to verify that we have a pure state
+    assert np.trace(noiseless_dm2**2).real == 1
```

Comment:
Better to use `np.isclose()` for floating-point comparison, to account for rounding error.

Reply:
Good point, done in [ff91d0a](https://github.com/CQCL/pytket-qiskit/pull/380/commits/ff91d0a4c9ad9116b4e63bfdd0d498591808eeec)